devito/ops/compiler.py

```diff
@@ -63,33 +62,77 @@ def _translate_ops(self, soname, ccode, hcode):
-            subprocess.run([translator, c_file.name], cwd=self.get_jit_dir())
+        # Calling OPS Translator
+        translator = '%s/../ops_translator/c/ops.py' % (self._ops_install_path)
+        translation = subprocess.run([translator, c_file.name], cwd=self.get_jit_dir())
+        if translation.returncode == 1:
+            raise ValueError("OPS Translation Error")
+
+
+class OPSSpecializedCompiler(OPSMetaCompiler):
+    def __init__(self, *args, **kwargs):
+        super(OPSSpecializedCompiler, self).__init__(*args, **kwargs)
+
+    def _cmdline(self, files, object=False):
+        if object:
+            ld_options = ['-c']
+            link = []
         else:
-            warning("Couldn't find OPS_INSTALL_PATH \
-                environment variable, please check your OPS installation")
+            ld_options = self.ldflags
+            link = ["-L%s" % ldir for ldir in self.library_dirs]
+            link.extend(["-l%s" % lib for lib in self.libraries])
+        return (
+            [self.cc]
+            + self.cflags
+            + ld_options
+            + ["-D%s" % define for define in self.defines]
+            + ["-U%s" % undefine for undefine in self.undefines]
+            + ["-I%s" % idir for idir in self.include_dirs]
+            + files
+            + link
+        )
+
+
+class OPSCompilerSEQ(OPSCompiler, OPSSpecializedCompiler):
+
+    def __init__(self, *args, **kwargs):
+        super(OPSCompilerCUDA, self).__init__(*args, **kwargs)
+
+    def _compile(self, soname):
+        pass
+
 
-    def _compile_cuda(self, soname):
+class OPSCompilerCUDA(OPSCompiler, OPSSpecializedCompiler):
+
+    def __init__(self, *args, **kwargs):
+        super(OPSCompilerCUDA, self).__init__(*args, **kwargs)
+
+        self._cuda_install_path = os.environ.get('CUDA_INSTALL_PATH')
+        if not self._cuda_install_path:
+            raise ValueError("Couldn't find CUDA_INSTALL_PATH a\
+                environment variable, please check your CUDA installation")
+
+        self._nv_arch = os.environ.get('NV_ARCH')
+        if not self._nv_arch:
+            raise ValueError("Select an NVIDIA device to compile in CUDA, e.g. \
+                NV_ARCH=Kepler")
+
+    def _compile(self, soname):
         # CUDA kernel compilation
         cuda_src = '%s/CUDA/%s_kernels.cu' % (self.get_jit_dir(), soname)
         cuda_target = '%s/%s_kernels_cu' % (self.get_jit_dir(), soname)
 
-        cuda_code = ""
         try:
             with open(cuda_src, 'r') as f:
                 cuda_code = f.read()
         except FileNotFoundError:
             raise ValueError("Couldn't find file: %s" % cuda_src)
 
-        cuda_device_compiler = CUDADeviceCompiler()
-        cuda_host_compiler = CudaHostCompiler()
-
         with warnings.catch_warnings():
             warnings.simplefilter('ignore')
 
             # Spinlock in case of MPI
+            cuda_device_compiler = CUDADeviceCompiler()
```

Comment:
It makes totally no sense to instantiate an object of the subclass in the superclass

this is just voodoo that Python lets you doing, but it simply is a software engineering error

Reply:
so is there a problem in line 45 of `ops/compiler.py` as well?
`OPSCompilerCUDA._compile(self, soname)` is called from `OPSCompiler`

Reply:
that is totally fine in object oriented programming (OOP). You should read something about polymorphism in OOP  (@maelso can you help him out please) 

the error I described above is that a superclass shouldn't know anything about what/who the subclasses are, while the converse is fine

Reply:
as far as i can see, CUDADeviceCompiler is not a subclass of OPSCompiler, but more like a 'sister' class. i'm not sure if that makes sense.
 
i'm uploading how i understand their relation:
![WhatsApp Image 2019-12-11 at 16 21 23](https://user-images.githubusercontent.com/17774383/70652993-6106ce80-1c32-11ea-9156-b94010758ce6.jpeg)

Reply:
You moved `cuda_device_compiler = CUDADeviceCompiler()` from **OPSCompilerCUDA** to **OPSCompiler**, that change solved the problem mentioned by Fabio. Now, there is an **association** between OPSCompiler and OPSCompilerCUDA.

Reply:
I'm not sure this association makes much sense. If you:

* move the cuda-specific stuff from OPSCompiler to OPSCompilerCuda (or any of its subclasses)
* move the compilation-specific stuff to OPSMetaCompiler (to be renamed as OPSCompiler)

wouldn't it become cleaner and simpler? What's your take, @maelso ?